pkg/core/klinebuilder.go

```diff
@@ -0,0 +1,221 @@
+package core
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/c9s/bbgo/pkg/types"
+	"github.com/sirupsen/logrus"
+)
+
+// kLineBuilder is a kline builder for a symbol
+// It will accumulate trades and build klines
+//
+//go:generate callbackgen -type kLineBuilder
+type kLineBuilder struct {
+	symbol        string
+	minInterval   types.Interval
+	intervals     map[types.Interval]struct{}
+	isBackTesting bool
+
+	kLineCallbacks       []func(kline types.KLine)
+	kLineClosedCallbacks []func(kline types.KLine)
+	logger               *logrus.Entry
+
+	//
+	mu      sync.Mutex
+	state   *kLineBuilderState
+	running bool
+
+	exchange types.Exchange
+}
+
+// NewKLineBuilder: constructor of kLineBuilder
+//   - symbol: symbol to trace on
+//   - minInterval: unit interval, related to your signal timeframe.
+//     All the supported intervals of the binding stream should be multiple of this interval.
+func NewKLineBuilder(symbol string, minInterval types.Interval, isBackTesting bool) *kLineBuilder {
+	logger := logrus.WithField("symbol", symbol)
+	// Interval.Duration() will panic on invalid value, ex: "-3m"
+	_ = minInterval.Duration()
+
+	return &kLineBuilder{
+		symbol:        symbol,
+		minInterval:   minInterval,
+		isBackTesting: isBackTesting,
+		logger:        logger,
+		intervals:     make(map[types.Interval]struct{}),
+		state:         NewKLineBuilderState(symbol),
+	}
+}
+
+func (kb *kLineBuilder) Subscribe(interval types.Interval) {
+	if interval.Duration()%kb.minInterval.Duration() != 0 {
+		err := fmt.Errorf(
+			"valid interval should be multiple of %s: %s is given",
+			kb.minInterval,
+			interval,
+		)
+		kb.logger.Error(err)
+		panic(err)
+	}
+	kb.intervals[interval] = struct{}{}
+}
+
+func (kb *kLineBuilder) BindStream(stream types.Stream) {
+	if kb.isBackTesting {
```

Comment:
can't handle backtest like this, would prefer pull out this logic to another layer